usr/bin/cylc

```diff
@@ -16,54 +16,156 @@
+#   version outside of the ROOT location
+#   - CYLC_HOME should point to a Cylc 8 venv, or a plain Cylc 7 directory. It
+#     must be set in .bashrc so that task jobs see it too
+# + Any of these settings in .bashrc must be replicated on job hosts.
+#
+##############################!!! EDIT ME !!!##################################
+# Centrally installed Cylc releases:
+CYLC_HOME_ROOT="${CYLC_HOME_ROOT:-/opt}"
 
-CYLC_HOME_ROOT=${CYLC_HOME_ROOT:-/opt} # !!! EDIT ME !!!
+# Note users can set CYLC_HOME_ROOT_ALT as well (see above), e.g.:
+# CYLC_HOME_ROOT_ALT=${HOME}/miniconda3/envs
+###############################################################################
 
-if [[ -z ${CYLC_HOME:-} ]]; then
-    if [[ -n ${CYLC_VERSION:-} && -d $CYLC_HOME_ROOT/cylc-$CYLC_VERSION ]]; then
-        CYLC_HOME=$CYLC_HOME_ROOT/cylc-$CYLC_VERSION
-    else
-        CYLC_HOME=$CYLC_HOME_ROOT/cylc
+# If CYLC_HOME is not defined try CYLC_HOME_ROOT/cylc-CYLC_VERSION
+if [[ -z "${CYLC_HOME}" ]]; then
+    if [[ -z ${CYLC_VERSION} ]]; then
+        echo 1>&2 "ERROR: CYLC_HOME or CYLC_VERSION must be defined"
+        exit 1
+    fi     
+    # Look for matching version in root locations
+    for ROOT in "${CYLC_HOME_ROOT}" "${CYLC_HOME_ROOT_ALT}"; do
+        if [[ -d "${ROOT}/cylc-${CYLC_VERSION}" ]]; then
+            CYLC_HOME="${ROOT}/cylc-${CYLC_VERSION}"
+            break
+        fi
+    done
+fi
+if [[ -z "${CYLC_HOME}" ]]; then
+    MSG="ERROR: cylc-$CYLC_VERSION not found in $CYLC_HOME_ROOT"
+    if [[ -n "${CYLC_HOME_ROOT_ALT}" ]]; then
+        MSG="${MSG} or ${CYLC_HOME_ROOT_ALT}"
     fi
+    echo 1>&2 "$MSG"
+    exit 1
 fi
-exec "$CYLC_HOME/bin/$(basename "$0")" "$@"
+
+# Find and activate the virtual env, or find the Cylc 7 bin directory.
+BASENAME_PREFIX=""
+if [[ -f "${CYLC_HOME}/bin/activate" ]]; then
+    # A Python venv or Conda pack installation
+    . "${CYLC_HOME}/bin/activate" || exit 1
+elif [[ -d "${CYLC_HOME}"/conda-meta && \
+        -f "${CYLC_HOME%/*/*}"/etc/profile.d/conda.sh ]]; then
+    # A normal Conda environment
+    . "${CYLC_HOME%/*/*}"/etc/profile.d/conda.sh
+    # Unset CONDA_SHLVL in case CONDA_* was inherited from the scheduler
+    # environment (this has no effect in a clean job submission shell). Doing
+    # this avoids an apparent bug where conda.py:reactivate() does not put the
+    # target environment in front of PATH. If bashrc has since prepended the
+    # wrapper location to PATH in the job environment this results in the
+    # wrapper invoking itself again (repeatedly... fork bomb).
+    unset CONDA_SHLVL
+    conda activate "${CYLC_HOME##*/}" || exit 1
+elif [[ -x "${CYLC_HOME}/bin/cylc" ]]; then
+    # A Cylc 7 installation
+    BASENAME_PREFIX="${CYLC_HOME}"/bin/
+else
+    echo 1>&2 "ERROR: cylc not found in ${CYLC_HOME}"
+    exit 1
+fi
+
+# Execute the command via the selected Cylc installation.
+exec "${BASENAME_PREFIX}${0##*/}" "$@"
```

Comment:
Why can't we stick with ```exec "${CYLC_HOME}/bin/${0##*/}" "$@"``` ?
That fixes the fork bomb issue (sorry - didn't notice this before).
As far as I can see this works in all cases.

Reply:
Damn it, you're right. I switched to relying on PATH to find `cylc` for the conda case because that's more in the spirit of virtual environments (find the environment and activate it, then rely on the "activated" PATH). It's probably still sensible to force `conda activate` to do the right thing with PATH just in case, but yes, going back to the explicit path/to/cylc would prevent the fork bomb either way. (Jeez, not my best effort, this PR :grimacing: !!)

Reply:
(Well, in my defense, if `conda activate` had done what it says on the tin (to PATH) my way wouldn't have caused a fork bomb either).